### PR TITLE
feat: allow replying to messages to bookmark

### LINKED ISSUE #689: Bookmark messages using the message reference (replying)
## Description
The ability to use the bookmark command without a message link argument to bookmark a message by replying to it. Like this:
image
(I already have the code since I was curious, but I need to make an issue so everything is done by the book)

## Reasoning
It's just a small QoL change that means people wouldn't need to copy the message link to bookmark the message, which is especially useful on Mobile/Android specifically since copying a message link is a pain because you have to go to the menu > share > copy link to get it.

## Proposed Implementation
Implementation is done by making the target_message argument optional, and then if it's unset trying to use the message.reference. To use the reference it tries to fetch the referenced message from the message cache, and if that's None it fetches it from the API.

## Would you like to implement this yourself?

- [x] I'd like to implement this feature myself
- [ ] Anyone can implement this feature

### PR DESCRIPTION
**Pull request by vcokltfre**
## Relevant Issues
Closes #689 

## Description
Users can now bookmark a message by replying to it, this is done by making the message argument optional, and if it's None take the message reference, if that is None raise an error to display command usage.

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] Read all the comments in this template?
- [x] Ensure there is an issue open, or link relevant discord discussions?
- [x] Read the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

### CODE REVIEW

File: bot/exts/evergreen/bookmark.py [hunk 1]
```diff
@@ -88,11 +89,16 @@ async def send_reaction_embed(
...
-        target_message: WrappedMessageConverter,
+        target_message: t.Optional[WrappedMessageConverter],
         *,
         title: str = "Bookmark"
     ) -> None:
         """Send the author a link to `target_message` via DMs."""
+        if not target_message:
+            if not ctx.message.reference:
+                raise commands.UserInputError("You must either provide a valid message to bookmark, or reply to one.")
+            target_message = await ctx.channel.fetch_message(ctx.message.reference.message_id)
```
Comment (ToxicKidz) [inline bot/exts/evergreen/bookmark.py:100]: Nitpick 🙂 
```suggestion
            target_message = await ctx.fetch_message(ctx.message.reference.message_id)
```
Then again, [`Message.resolved`](https://discordpy.readthedocs.io/en/latest/api.html#discord.MessageReference.resolved) exists, but I'm not sure how reliable that is.

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

File: bot/exts/evergreen/bookmark.py (hunk 1 above)
Comment (vcokltfre) [reply bot/exts/evergreen/bookmark.py:100]: Actually message.reference.resolved doesnt exist because lance uses 1.5.1 not 1.6

**Comment by vcokltfre** [comment]
Looks like
image

**Review by JakeM-Git — APPROVED** [review]

**Review by ToxicKidz — APPROVED** [review]

File: bot/exts/evergreen/bookmark.py [hunk 2]
```diff
@@ -88,11 +89,16 @@ async def send_reaction_embed(
...
-        target_message: WrappedMessageConverter,
+        target_message: t.Optional[WrappedMessageConverter],
         *,
         title: str = "Bookmark"
     ) -> None:
         """Send the author a link to `target_message` via DMs."""
+        if not target_message:
+            if not ctx.message.reference:
+                raise commands.UserInputError("You must either provide a valid message to bookmark, or reply to one.")
+            target_message = await ctx.fetch_message(ctx.message.reference.message_id)
```
Comment (ChrisLovering) [inline bot/exts/evergreen/bookmark.py:100]: Could we try to get this from the cache before making an API request?

Comment (vcokltfre) [reply bot/exts/evergreen/bookmark.py:100]: Ideally if lance is using discord.py 1.6.0 or above we can just use the resolved reference so we never need to fetch from the API

Comment (vcokltfre) [reply bot/exts/evergreen/bookmark.py:100]: currently it's 1.5.1

**Review by ChrisLovering — APPROVED** [review]
Tested locally, all good 👍

**Pull request merged**